### ISSUE TITLE
Calling an external setter function from the setter block results in a stack overflow and a hard crash

### ISSUE THREAD

**Issue by SaracenOne** [godotengine/godot#48437]
**Godot version:**
8abd50359ba7b5af38a4ceedba287b4459e5f925

**OS/device including version:**
Windows 10 

**Issue description:**
When using a setter, if you attempt to use an external function as part of the setter, the engine will suffer a stack overflow and a hard engine crash since there's no protection against cyclic recursion.

**Steps to reproduce:**
Simply run the example project to observe the crash

**Minimal reproduction project:**
[setter_test.zip](https://github.com/godotengine/godot/files/6418725/setter_test.zip)

**Comment by lyuma** [comment]
this new set get syntax is a newbie trap.

Forgot a comma after `get=something`? Syntax error but you don't know why
Added a comma after `get: return foo`? Syntax error but you don't know why
Did `set(x): setter(x)`? Infinite loop
Did `set(x): setter`? Setter doesn't do anything (it's just evaluating a function reference and doing nothing with it.
Mix/match inline `get:` with `set=`? Syntax error

It's also not obvious that `get: return prop` can be omitted if it's the default getter.

I recommend doing
```
@export var foo: int:
    set = my_setter_func
```
but again make sure to use `=` because `:` will not give a syntax error but it won't actually work.

I think this should be reviewed seriously and figure out how to give warnings or errors to alleviate the newbie traps here.

Don't just take my word for it. Here is the Discord conversation: https://discord.com/channels/212250894228652034/908062856459743262/936262756896735252
They went through all of the pitfalls I listed above, one by one, and I was forced to clarify all the syntactic nuances to fix it.

**Comment by Proggle** [comment]
Yeah the new syntax is a real nightmare to use.

I especially dislike how it's inconsistent, where

```
var health: int = max_health:
	set(value):
		health = clamp(value, 0, max_health)
```

will work correctly, but 

```
var health: int = max_health:
	set(value):
		health = set_health(value)

func set_health(value): 
	health = clamp(value, 0, max_health)
```

will not, because the engine ignores recursion in the first layer of the stack within the setter, but allows for recursion on any deeper layers.

Why not just have the setter ignore itself on all layers, instead of just the first one?  It would make the setters work more consistently and prevent them from being a massive newbie trap.

And while I'm sure there's somebody living in a cave somewhere who has some hyperspecific edge-case optimization that requires a recursive-function-as-setter, it seems like they could do something like

```
var health: int = max_health:
	set(value):
		health = set_health_recursive(value)

func set_health_recursive(value):
	#do some complex optimization
	# godot 3 script used "health = value" as the call to start recursing
	# godot 4 script uses a direct call to this function to recurse instead:
	set_health_recursive(value)
```
And just explicitly invoke the function rather than getting redirected to it as a setter.

**Comment by Calinou** [comment]
cc @vnen

**Comment by vnen** [comment]
I believe the only thing that could be said is a bug here is not having a check for infinite recursion. Even that is tricky because we need to find a value that is not too big so it doesn't crash before it's reached, nor too small so to hinder the usage of proper recursive functions.

We could potentially check if the function calls itself unconditionally, but that wouldn't work for all the cases.

Regarding the syntax for property, we can try to tweak if there are better suggestions. But setting a variable that has a setter will always call the setter, so that it is consistent (the fact that is called from the setter itself shouldn't matter as we don't want a function to behave differently depending from where it's being called).

**Comment by lyuma** [comment]
This issue is not fixed. Reopening.

**Comment by lyuma** [comment]
Yes, the fact that infinite recursion causes hard crash is a problem which is what this issue was filed about.

But also, it's really easy to accidentally cause inifnite recursion (hard crash or no) by calling a function in the setter with no warning. I think the lack of a warning or detection for this ought to be its own issue since it might be something that people new to Godot might run into.

**Comment by lyuma** [comment]
I think the issue speaks for itself so a stack trace isn't needed, but you can see here:

```
 	godot.windows.tools.64.exe!GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) Line 1547	C++
 	godot.windows.tools.64.exe!GDScriptInstance::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 1567	C++
 	godot.windows.tools.64.exe!Object::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 662	C++
 	godot.windows.tools.64.exe!Variant::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error) Line 1048	C++
 	godot.windows.tools.64.exe!GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) Line 1547	C++
 	godot.windows.tools.64.exe!GDScriptInstance::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 1567	C++
 	godot.windows.tools.64.exe!Object::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 662	C++
 	godot.windows.tools.64.exe!Variant::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error) Line 1048	C++
 	godot.windows.tools.64.exe!GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) Line 1547	C++
 	godot.windows.tools.64.exe!GDScriptInstance::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 1567	C++
 	godot.windows.tools.64.exe!Object::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 662	C++
 	godot.windows.tools.64.exe!Variant::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error) Line 1048	C++
 	godot.windows.tools.64.exe!GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) Line 1547	C++
 	godot.windows.tools.64.exe!GDScriptInstance::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 1567	C++
 	godot.windows.tools.64.exe!Node::_gdvirtual__ready_call<0>() Line 241	C++
 	godot.windows.tools.64.exe!Node::_notification(int p_notification) Line 155	C++
 	godot.windows.tools.64.exe!Node::_notificationv(int p_notification, bool p_reversed) Line 45	C++
 	godot.windows.tools.64.exe!Object::notification(int p_notification, bool p_reversed) Line 743	C++
 	godot.windows.tools.64.exe!Node::_propagate_ready() Line 188	C++
 	godot.windows.tools.64.exe!Node::_propagate_ready() Line 180	C++
 	godot.windows.tools.64.exe!Node::_set_tree(SceneTree * p_tree) Line 2588	C++
 	godot.windows.tools.64.exe!SceneTree::initialize() Line 400	C++
 	godot.windows.tools.64.exe!OS_Windows::run() Line 775	C++
 	godot.windows.tools.64.exe!widechar_main(int argc, wchar_t * * argv) Line 179	C++
 	godot.windows.tools.64.exe!_main() Line 201	C++
 	godot.windows.tools.64.exe!main(int argc, char * * argv) Line 215	C++
```

**Comment by cdemirer** [comment]
In that case the issue seems to be even more complicated.

Besides the issue only happening with `target=debug`, on my system I never get a hard crash, instead the game freezes. But people report that it can lead to a crash as well.

I guess we need to find out why the stack overflow check is not happening when compiled with `target=debug`.

> But also, it's really easy to accidentally cause inifnite recursion (hard crash or no) by calling a function in the setter with no warning. I think the lack of a warning or detection for this ought to be its own issue since it might be something that people new to Godot might run into.

This can't always be detected compile-time. I think the existing infinite recursion warning would suffice (the way it currently works with `target=release_debug`).

**Comment by bruvzg** [comment]
> Somehow reduce the stack frame size for `target=debug` builds

Or increase stack size for all `debug` builds (like we currently do for the builds with enabled sanitizers):

https://github.com/godotengine/godot/blob/42537daeb1819b72a700acfe7ab6486b1a5d45d8/platform/linuxbsd/godot_linuxbsd.cpp#L44-L48

**Comment by lyuma** [comment]
What about release builds? Wouldn't those also need to have more stack space? Or is there some optimization in release that allows more stack frames to fit.

What if we reduce gdscript to 256 frames? That would be enough for most scripts and avoid this crash. The limit could be soft-coded so devs who want more can raise it.

**Comment by vnen** [comment]
This was brought attention during our GDScript meeting. It seems we need some investigation to understand the proper causes of this. For instance, is the issue that we are using too much stack space or just that the limit is too high?

@cdemirer also suggested increase the stack size in debug builds to avoid such errors as well. This might be good, but we need to make sure that what works on the editor works on release too.

Maybe we need to check remaining stack size instead of recursion depth to make sure it works.

In short, there's a problem but we don't have consensus on the solution yet.

**Comment by cdemirer** [comment]
I did some further testing. The stack usage of `GDSCriptFunction::call` highly depends on how exactly Godot is compiled (mostly optimization flags). For example, if you compile with `target=release_debug use_llvm=yes` but also add `optimize=none`, then even a GDScript call depth of ~25 will be too much for the default limits because `GDSCriptFunction::call` uses more than 300 kB of stack area...

---

I used `__builtin_frame_address` to estimate the stack usage of `GDScriptFunction::call`.
```cpp
uint64_t estimation = (uint64_t)(__builtin_frame_address(1)) - (uint64_t)(__builtin_frame_address(0));
__print_line(vformat("Stack frame size: %d\n", estimation));
```
I added this in `Variant::callp` (which is called by `GDSCriptFunction::call`) and ran with a simple recursive GDScript function. Here's some data:

> **Note**
> The content/complexity of the GDScript code doesn't seem to affect the results (at least not by much).

> **Note**
> Tested with clean builds for each case. I have to use `use_static_cpp=no` in order to compile on my OS (#\58374). I assume it doesn't affect the results much. For brevity I didn't include it in the commands below.

* `scons platform=linuxbsd use_llvm=no target=release_debug` (`optimize=speed`) (`-O2`)
Stack frame size: `1296`

* `scons platform=linuxbsd use_llvm=yes target=release_debug` (`optimize=speed`) (`-O2`)
Stack frame size: `2096`

---

* `scons platform=linuxbsd use_llvm=no target=release_debug optimize=size` (`-Os`)
Stack frame size: `1152`

* `scons platform=linuxbsd use_llvm=yes target=release_debug optimize=size` (`-Os`)
Stack frame size: `2064`

---

* `scons platform=linuxbsd use_llvm=no target=release_debug optimize=none`
Stack frame size: `150432`

* `scons platform=linuxbsd use_llvm=yes target=release_debug optimize=none`
Stack frame size: `331600`

---

* `scons platform=linuxbsd use_llvm=no target=debug`
Stack frame size: `7584`

* `scons platform=linuxbsd use_llvm=yes target=debug`
Stack frame size: `21712`

  * Why is `target=debug` better than `target=release_debug optimize=none`?

**Comment by lostminds** [comment]
I ran into this issue trying to use an external function in the setter as mentioned above. To make this a little more urgent, in my case it was even worse since I ran into the issue in a @tool script, meaning it crashed the editor. And since the editor open scene had the node in it the editor would crash again instantly on trying to re-open the project. In the end I had to find the script and edit it with an external text editor and change the script from there. Luckily, by that time I had figured out it was the external function in the setter, otherwise I don't know how I would have dealt with the problem.

**Comment by adamscott** [comment]
Could we check, on call, the stack size versus the allowed resource limits? This could theoretically make the allowed recursion limit match the current system/build type?

**Comment by vnen** [comment]
I did a bit of research and it seems there's not really a way to find out the stack usage, at least no in a portable way (I'm not sure if all platforms even have a way to do it). So essentially we can't tell we're about to run out of stack to abort the call.

The best we can do is having a limit in the call depth. We do have that now but only for the debugger, if you're running without it, even with a debug build, nothing will check this. I don't think having this configurable is a good thing, the users might increase it thinking it's gonna allow deeper recursion, but it will only going to crash.

The hard part is figuring out the magic number that is not too little that could be reached by regular deep/recursive calls and not too big that would still crash often (also considering multiple platforms as well as debug vs. release). Testing the MRP from this issue with a debug build I saw it reaching a bit over 300 depth before crashing. It's much less than the default of 1024, which makes me think it's not an appropriate number, especially given is a very simple function.

I'll try to test on other platforms w/ debug and release to see if there's a good candidate number to use for this. The check for depth should happen even in release builds to avoid crashing.

**Comment by vnen** [comment]
A bit more testing and I see that 1024 may be fine. An optimized editor build got over 3400 calls deep with this test, a release build got over 5000 calls deep. The only thing is that this limit should probably be baked into the binary and not configurable. Even so, there is still a chance you use too much stack in your functions and this limit is reduced (although unlikely, given GDScript stack usage for each call is essentially an array of Variants and unless you have thousands of local variables you won't use a lot of stack slots).
